### ISSUE TITLE
[FR]: Admin SDK should have a way to generate a session cookie without an ID token, or SDK should have a way to generate ID tokens

### ISSUE THREAD

**Issue by Ranguna** [firebase/firebase-admin-node#945]
**Describe the solution you'd like**
Currently, to generate a session cookie, an ID token is needed. This ID token can only be generated by the client SDK through a "sign in" method like `signInWithEmailAndPassword` or `signInWithCustomToken`, then this ID can be sent over to a backend (ex: a cloud function) and there a session cookie is generated.
The problem is that we need frontend logic for this which is quite cumbersome to do if I want to manage cookies server side and not leak any logic to the frontend (expect calling a single endpoint with the user credentials), as well as not dealing with cross site request forgery by using the `sameSite` cookie flag.

To do this, I would like a way to generate a session cookie without having an ID token, for example:
`user.generateSessionCookie(sessionCookieOptions)`
or
`admin.auth().generateSessionCookie(uid/UserRecord/cutomToken, sessionCookieOptions)`

**Describe alternatives you've considered**
One alternative that I'm going to have to take is to call the endpoint described in this link: https://firebase.google.com/docs/reference/rest/auth/#section-verify-custom-token. To exchange a custom token with an ID token and use this ID token to generate a cookie session that can last up to two weeks.
Having to do this:
1. get user (login with credentials/create new user);
2. generate custom token;
3. call endpoint to exchange for an ID token;
4. generate session cookie with ID token;
5. Use session cookie.

Is quite messy, not to mention that it's strange that we don't have a good way to do this yet.
I also thought about using the client SDK on the server side, but the login methods are rate limited and I think I'd hit that limit quickly if I call it from a single or a few dozen server instances.

**Comment by bojeil-google** [comment]
Hey @Ranguna, please file a feature request via [official channels](https://firebase.google.com/support). This is a pretty big design change and as this is a client SDK repo, the request should go through official channels.
Here are my thoughts on this:

1. Firebase is optimized for client side serverless rendering. Flows like phone auth, federated sign in are currently rendered client side (Game Center sign-in, Play Games sign in, Apple sign in, etc). In order to support this, we need to provide server side APIs that can validate credentials passed back to a server but even then, there will still be a strong dependency on client side credentials and assertions (eg. reCAPTCHA dependencies, etc).
2. Currently the admin sdk does not support authentication operations. This has to be supported first before we can support the requested feature.
3. The Admin SDKs are not used for minting standalone cookies (which seems to be the request here). If you want to mint standalone cookies, you may as well use another service for this. For example, express / passport is a good service for that in Node.js.

Ideally in order to support your request, we need to be able to pass the raw user credentials server side and provide an API to exchange the credentials for a session cookie directly. It is unlikely, we will just mint session cookies without some credential as you are requesting. This would essentially make us a JWT utility.

**Comment by Ranguna** [comment]
> please file a feature request via official channels.
Will do, thank you!

> In order to support this, we need to provide server side APIs that can validate credentials passed back to a server 
Makes sense, but this is not true for the case of email/password accounts and I guess it would be a good first step for this.

> The Admin SDKs are not used for minting standalone cookies
Then I'm a little confused with the use of the (createsessioncookie)[https://firebase.google.com/docs/reference/admin/node/admin.auth.Auth#createsessioncookie] method.

> For example, express / passport is a good service for that in Node.js.
Indeed, using `passport` would be a good option here.

> mint session cookies without some credential as you are requesting
My original request: I want to manage cookies server side and not leak any logic to the frontend (expect calling a single endpoint with the user credentials).

**Comment by bojeil-google** [comment]
`createSessionCookie` provides a mechanism to generate session cookies (for longer server side sessions) given an ID token, because short lived ID tokens may not be convenient for traditional applications that rely on server side cookies.

What you were asking is a standalone cookie generator: `user.generateSessionCookie(sessionCookieOptions)`. This wouldn't even be a Firebase feature. It is basically a standalone mechanism to generate session cookies. This wouldn't validate credentials. It would just issue cookies based on the options you provide.

A more feasible option is to provide a session cookie generation given a user credential, bypassing the ID token step.
Basically instead of: user credential -> id token -> session cookie. You would have: user credential -> session cookie.

I think that is more feasible, but it is still a major change in the scope of what the Admin SDKs cover.

**Comment by Ranguna** [comment]
> A more feasible option is to provide a session cookie generation given a user credential, bypassing the ID token step.

I see the confusion. Yes that would work perfectly!

Right now I'm using the [signInWithPassword](https://firebase.google.com/docs/reference/rest/auth/#section-sign-in-email-password) endpoint from firebase rest API to signin the user on the server side and exchange the id token for a session cookie.

**Comment by Ranguna** [comment]
A request was sent though the official channels. I'll close this issue.

**Comment by rromanchuk** [comment]
@bojeil-google sorry to bug you.  i currently use rest APIs only, and dont have an admin SDK, i'm curious about a couple of things under the hood as i may be over complicating things. 

I'm already minting my own jwt tokens in exchange for valid idTokens via server side using admin rest api, being used in a mobile client/server context.  I'm bringing some of this functionality to traditional web and would prefer to have the client login via REST post, token is validated, then session set via server.  are `createSessionCookie` and `verifySessionCookie` doing something i'm not already doing?  I found this https://cloud.google.com/identity-platform/docs/reference/rest/v1/projects/createSessionCookie  passing the users signed token with service account but the API isnt behaving the same way as the documented/branded firebase rest apis.

All the examples use the SDK and implementation details are obfuscated away but hint to backing rest, so just wondering if it's something unique, or is it using standard apis, just exchanging refresh tokens. I'm going to dig into some of the other sdks and just grep for endpoints

**Comment by rromanchuk** [comment]
oh dang, it is hitting a named endpoint

https://github.com/firebase/firebase-admin-node/blob/d961c3f705a8259762a796ac4f4d6a6dd0992eb1/test/unit/auth/auth-api-request.spec.ts#L113

**Comment by rromanchuk** [comment]
ok this helps a lot actually.  Posting for future SEO searches 

https://github.com/firebase/firebase-admin-python/blob/5e88d927757e61023e274cd8b00270d899fdf513/firebase_admin/_token_gen.py#L157

**Comment by bojeil-google** [comment]
All endpoints used are standard REST APIs. More documentation available [here](https://cloud.google.com/identity-platform/docs/reference/rest).
